File: proposed-projects/round-trip-experiment/reverse/20260909T182714Z_2643e4be/20260910T100600Z_9e4cc6a6/ported_implementation/gbnf/grammar_graph/pointers.py
from __future__ import annotations

from typing import Dict, Iterator

from .graph_pointer import GraphPointer


class Pointers:
    def __init__(self, *pointers: GraphPointer) -> None:
        self._pointers: Dict[str, GraphPointer] = {}
        for pointer in pointers:
            self.add(pointer)

    def add(self, pointer: GraphPointer) -> None:
        self._pointers[pointer.id] = pointer

    def __iter__(self) -> Iterator[GraphPointer]:
        return iter(list(self._pointers.values()))

    def __len__(self) -> int:
        return len(self._pointers)

    @property
    def size(self) -> int:
        return len(self._pointers)

    def __str__(self) -> str:
        return f"<Pointers [{', '.join(str(p.node.rule) for p in self)}]>"

    def __repr__(self) -> str:
        return self.__str__()

File: packages/gbnf-experiment/data/20260909T151143Z_170b737d/ported_implementation/src/grammar_graph/pointers.ts
import type { GraphPointer } from "./graph_pointer.ts";

export class Pointers {
  private __pointers__: Map<string, GraphPointer>;

  constructor(...pointers: GraphPointer[]) {
    this.__pointers__ = new Map();
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  toString(): string {
    return `<Pointers ${[...this].map((p) => String(p.node.rule)).join(", ")}>`;
  }

  add(pointer: GraphPointer): void {
    this.__pointers__.set(pointer.id, pointer);
  }

  *[Symbol.iterator](): Generator<GraphPointer> {
    yield* this.__pointers__.values();
  }

  get length(): number {
    return this.__pointers__.size;
  }
}
